src/puppetlabs/puppetdb/utils.clj

```diff
@@ -264,6 +264,55 @@
       (apply assoc m new-kvs)
       m)))
 
+(defn str-last
+  [s]
+  (if (= s "")
+    nil
+    (some-> s (.charAt (dec (.length s))))))
+
+(defn maybe-strip-escaped-quotes
+  [s]
+  (if (and (> (count s) 1) (= \" (first s) (str-last s)))
+    (subs s 1 (dec (count s)))
+    s))
+
+(defn parse-matchfields
+  [s]
+  (string/replace s #"match\((\".*\")\)" "$1"))
+
+(defn parse-indexing
+  [s]
+  (string/replace s #"\[(\d+)\]" ".$1"))
+
+(defn split-indexing
+  [path]
+  (flatten
+    (for [s path]
+      (if (re-find #"\[\d+\]$" s)
+        (-> s
+            (string/split #"(?=\[\d+\]$)")
+            (update 1 #(Integer/parseInt (subs % 1 (dec (count %))))))
+        s))))
+
+(defn handle-quoted-thing
+  [v]
+  (loop [result []
+         [s & splits] v]
+    (if (and (= \" (str-last s)) (not (= (count s) 1)))
+      [(string/join "." (conj result s)) splits]
+      (recur (conj result s) splits))))
+
+(defn dotted-query->fact-path
+  [string]
+  (loop [[s & splits :as v] (string/split string #"\.")
+         result []]
+    (if (nil? s)
+      result
+      (if (and (= \" (first s)) (or (= (count s) 1) (not (= \" (str-last s)))))
```

Comment:
I think `clojure.string/starts-with?` and `clojure.string/ends-with?` would be easier to read here